### DISCUSSION TITLE
[help][functions] Calling HTTP Functions from Android is failing [Error: Response is not valid JSON object.]

### DISCUSSION THREAD

**Discussion by nertilzhuri** [invertase/react-native-firebase#5439] [Q&A] [answered]
Hello Everyone,

I am using firebase functions on my mobile app, but somehow can't seem to connect to any firebase function.

My firebase function looks like the following

```javascript
exports.testFunction = functions.region("europe-west3").https.onCall((data, context) => {
    console.log("HERE")
    return {
        success: "Yes",
        message: "HERE"
    }
})
```

On react-native, I have the following call:

```javascript
functions()
      .httpsCallable('testFunction')()
      .then(response => {
        console.log("Function response", response)
      })
      .catch(err => {
        console.log("Function failed", err)
      });
```

To connect to the `europe-west3` server I use the following:

```javascript
import firebase from "@react-native-firebase/app"
firebase.app().functions('europe-west3')
```

All my tries lead to the same error:

```
Function failed [Error: Response is not valid JSON object.]
```

Additional note: The function is working perfectly using the web SDK on web browsers.

**Any help is appreciated**

My rnfirebase versions from package.json are:

```json
"@react-native-firebase/app": "^12.1.0",
"@react-native-firebase/auth": "^12.1.0",
"@react-native-firebase/crashlytics": "^12.1.0",
"@react-native-firebase/database": "^12.1.0",
"@react-native-firebase/functions": "^12.1.0",
"@react-native-firebase/storage": "^12.1.0",
"react": "16.13.1",
"react-native": "0.63.3",
```

And I'm testing in an Android physical device (Xiaomi mi 8 with Android 10)

Let me know if you need any other information and **thank you for your time.**

**Comment by mikehardy** [comment, marked as answer]
This is a function problem. Not sure why the web SDK doesn't flag it, but your function is not returning JSON in the way we expect. 
In particuar, double-quotes are not actually valid JSON are they? I wonder if it works with single quotes.
Here's a testing library of functions we know work, from the related FlutterFire project

https://github.com/FirebaseExtended/flutterfire/blob/master/.github/workflows/scripts/functions/src/index.ts

**Reply by nertilzhuri** [reply]
Thank you for your time and response.

I tried with single quotes, I even tried wrapping the return with a JSON.stringify, none of it worked unfortunately.

I intentionally left a console.log on the function, which is triggered and the message is logged when the function is called from web sdk, but there is no log when I try from react-native-firebase.

Am I using the location setting correctly? Maybe I'm missing something there?

**Reply by mikehardy** [reply]
I'm really not sure, sorry. Might work if you format the JSON object to be more what Firebase expects though, specifically with a "data" element in the JSON you return

https://duckduckgo.com/?q=firebase+functions+Error%3A+Response+is+not+valid+JSON+object

https://firebase.google.com/docs/functions/callable-reference#response_body

> f the response is not a JSON object, or **does not contain data or error**, the client SDK should treat the request as failed with Google error code INTERNAL (13).

emphasis mine

**Reply by nertilzhuri** [reply]
Thank you for your help. 

I just found out that calling functions of another region is not working; when I removed the region from the function, everything was working as it should on the app. 

My assumption is that I am not setting up the region as I should using `react-native-firebase`. I am using the following code:
```javascript
import firebase from "@react-native-firebase/app"
firebase.app().functions('europe-west3')
```

If the code snippet above is correct, maybe I should open an issue regarding this? 
As well as having a proper error message in case function calling fails as in my case; I think the `Response is not a valid JSON` comes through firebase function errors.

What do you think?

**Reply by mikehardy** [reply]
This is how you use a custom region:

https://github.com/invertase/react-native-firebase/blob/46556de7ce17ef1d55b52222a907732353addde6/packages/functions/e2e/functions.e2e.js#L37-L53

It appears to be documented correctly https://rnfirebase.io/reference/functions ? (example 3)

If there is something that can be done to improve the error message that would be a great PR. You can reach right in to node_modules and modify the .java or .m files to log out the error message direct from the underlying SDK to see exactly what it is - if there is some information there that we are not passing (thus resulting in a worse error message) then fixing that would be great - if the error message is generic though then it would be a good upstream issue in either firebase-ios-sdk or firebase-android-sdk repository to make a more specific error message

**Reply by nertilzhuri** [reply]
You are right, its my bad, I was setting the region through `@react-native-firebase/app` but calling functions through `react-native-firebase/functions`.

To anyone having similar problems, this is the proper way of calling region based functions:

```javascript
import firebase from "@react-native-firebase/app"

const region = 'europe-west1'; 
const functionsForRegion = firebase.app().functions(region);

...

functionsForRegion.httpsCallable("....")  
```

Thank you very much for your help, much appreciated!
